### DISCUSSION TITLE
router.query not correctly populated using next-routes and no getInitialProps

### DISCUSSION THREAD

**Discussion by inside** [vercel/next.js#45857] [Help]
Hi,

We're trying to get rid of getInitialProps where it's possible to benefit from static optimization. We're still using next-routes (trying to move away from that either but in a next step) and I've noticed router.query isn't correctly populated (it's missing dynamic segments keys and values) on a page refresh. It works ok on a client side navigation.

Here's a repo to illustrate what I'm facing: https://github.com/inside/router-query

Do you guys have any solution to have all router.query parameters?
Thanks in advance

**Comment by icyJoseph** [comment]
Could it have to do with:

> query: Object - The query string parsed to an object, including [dynamic route](https://nextjs.org/docs/routing/dynamic-routes) parameters. It will be an empty object during prerendering if the page doesn't use [Server-side Rendering](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props). Defaults to {}

When Next.js does static rendering, the query parameter is initially empty, and as soon as hydration is completed, Next.js triggers a re-render with the actual query information.

Why is that? Because during pre-rendering, there's no browser, nor URL, no queries, so the page is pre-rendered using an empty query, so that you can have a starting template so-to-speak.

More on that:

> Pages that are statically optimized by [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization) will be hydrated without their route parameters provided, i.e query will be an empty object ({}).


I'll review the repo in a minute, but this could be the reason.

**Reply by inside** [reply]
Hi Joseph! I thought so in the first place too, but even when `router.isReady` is set to true. `router.query` is still missing parameters

**Reply by icyJoseph** [reply]
Oh, and it never fills in? That's no bueno...

**Comment by aaronpurewal** [comment]
Same issue here -- following

**Comment by icyJoseph** [comment]
I've tried to reproduce this and I just can't, could you show me how are you starting your server? 

https://user-images.githubusercontent.com/21013447/219021522-2767600f-bb97-411e-886f-3cd1955b5988.mov


I did this:

1. Open the repo with https://stackblitz.com/github/inside/router-query?file=README.md
2. After it installs dependencies, kill the running command, and do `npm run dev`, which calls nodemon

![Screenshot 2023-02-15 at 13 00 48](https://user-images.githubusercontent.com/21013447/219021892-f7c86ec0-e6f9-419b-a05e-22e23cc6c343.png)

3. Then add to the URL (on the right side) to `/a?slug=foo`, and hit enter

![Screenshot 2023-02-15 at 13 01 46](https://user-images.githubusercontent.com/21013447/219022062-8fa4454f-5b79-431d-be52-9c8383fc1e22.png)

4. I see the logs for the query:

![Screenshot 2023-02-15 at 13 02 10](https://user-images.githubusercontent.com/21013447/219022154-52bfccaa-6a45-40ec-8149-6acf8e60cdb9.png)

Could any of you tell me how to reproduce it?

**Reply by icyJoseph** [reply]
Aha, I see the issue if I use `next dev` directly, that is without the custom server!

**Reply by inside** [reply]
With `npm run dev` If you hit directly this page: http://localhost:3000/a/hello-world then the query is empty. It works ok though by first hitting the index page http://localhost:3000/ and then clicking on the "a" link

**Reply by icyJoseph** [reply]
I see, at first I thought you needed a page called `pages/a/[slug].tsx`, but I reckon the whole point of this library was to not do that by defining:

```js
const routes = require('next-routes')

module.exports = routes()
  .add(
    'a', // route name
    '/a/:slug', // route pattern
    'a' // page
  )
  .add('b', '/b/:slug', 'b')
```

I am also not sure on how wise it is to spend time investigating an issue with a library that deprecated itself 5 years ago, but the problem must be in the custom server + the `next-routes` matcher and `getRequestHandler`, but like I said, not sure I wanna invest time hunting a problem there, since the library was abandoned 5 years ago.

**Reply by inside** [reply]
Yes I'm looking for a quick win, since we're also planning on using nextjs router but it will take us months...

**Reply by icyJoseph** [reply]
I understand, it's just that it looks like it is fundamentally broken, because requests to the server do not even contain the query, when Next.js renders `a/:slug` it makes a request for `a/%5Bslug%5D`, and that tells the server that this is a dynamic route, and it is rendered appropriately. 

I don't know when or how it broke for `next-routes`, but if the above is correct, then all bets are off I think. The server will never know about `a/:slug` being requested, it will only see `a`, so the first frame won't have the dynamic parameter, slug.

On SPA transitions into the page the Client side Router takes over, and that sees the query through next/link, so it looks like it works fine.
